File: go/az-list-secrets.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/exec"
	"strings"
	"sync"
)

type ListingEntry struct {
  Id string
}

type ShowEntry struct {
  Value string
}

func listSecrets(vaultName string) []ListingEntry {
	cmd := exec.Command("az", "keyvault", "secret", "list", "--vault-name", vaultName)

  out, err := cmd.CombinedOutput()

	if err != nil {
		log.Fatalf("Listing failed with %s:\n%s\n", err, out)
	}

	fmt.Printf("Listing:\n%s\n", string(out))

	var entries []ListingEntry

	json.Unmarshal(out, &entries)

	return entries
}

func printSecret(vaultName string, secretId string) {
	cmd := exec.Command("az", "keyvault", "secret", "show", "--id", secretId, "--vault-name", vaultName)

	out, err := cmd.CombinedOutput()

	if err != nil {
		log.Fatalf("Show failed with %s:\n%s\n", err, out)
	}

	var showEntry ShowEntry

	json.Unmarshal(out, &showEntry)

	splitId := strings.Split(secretId, "/")
	name := splitId[len(splitId) - 1]

	fmt.Printf("Name  | %s\nValue | %s\n\n", name, showEntry.Value)
}

func main() {
	vaultName := os.Args[1]

	fmt.Printf("Connecting to Azure Key Vault: %s\n\n", vaultName)

	listingEntries := listSecrets(vaultName)

	fmt.Print("Secrets:\n\n")

	var wg sync.WaitGroup
	wg.Add(len(listingEntries))

	for i := 0; i < len(listingEntries); i++ {
		go func(i int) {
			defer wg.Done()
			entry := listingEntries[i]
			printSecret(vaultName, entry.Id)
    }(i)

	}

	wg.Wait()
}

File: python/az_list_secrets.py
#!/usr/bin/env python
import sys
import subprocess
import json

vault_name = sys.argv[1]

print("Connecting to Azure Key Vault: " + vault_name)

listing = subprocess.check_output(["az", "keyvault", "secret", "list", "--vault-name", vault_name])

print("\nListing:")

print(listing)

entries = json.loads(listing)

print("Secrets:\n")

for entry in entries:
  secret_id = entry['id']

  raw_secret = subprocess.check_output(["az", "keyvault", "secret", "show", "--id", secret_id, "--vault-name", vault_name])

  secret = json.loads(raw_secret)

  secret_name = secret_id.split("/")[-1]
  secret_value = secret['value']

  print("Name  | " + secret_name)
  print("Value | " + secret_value + "\n")
